pkg/block/indexheader/parallel_bucket.go

```diff
@@ -0,0 +1,251 @@
+// Copyright (c) The Thanos Authors.
+// Licensed under the Apache License 2.0.
+
+package indexheader
+
+import (
+	"bufio"
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+
+	"github.com/google/uuid"
+	"github.com/pkg/errors"
+	"github.com/prometheus/prometheus/tsdb/fileutil"
+	"github.com/thanos-io/objstore"
+	"github.com/thanos-io/thanos/pkg/runutil"
+	"golang.org/x/sync/errgroup"
+)
+
+// partitionSize is used for splitting range reads.
+const partitionSize = 16 * 1024 * 1024 // 16 MiB
+
+type parallelBucketReader struct {
+	objstore.BucketReader
+	tmpDir        string
+	partitionSize int64
+}
+
+func WrapWithParallel(b objstore.BucketReader, tmpDir string) objstore.BucketReader {
+	return &parallelBucketReader{
+		BucketReader:  b,
+		tmpDir:        tmpDir,
+		partitionSize: partitionSize,
+	}
+}
+
+// GetRange reads the range in parallel
+func (b *parallelBucketReader) GetRange(ctx context.Context, name string, off int64, length int64) (io.ReadCloser, error) {
+	partFilePrefix := uuid.New().String()
+	g := errgroup.Group{}
```

Comment:
I suggest we use `errgroup.WithContext(ctx)` and use the returned context when downloading individual parts. This way if one part fails, we would cancel downloading the rest.